# 2. Task_4.8

## Content :
### Step 1 : Import libraries and dataframe
### Step 2 : Aggregate mean of Order Number by Department
### Step 3 : How do the results for the entire dataframe differ from those of the subset?
### Step 4 : Create a Loyalty Flag
### Step 5 : Analyze Spending Habits by Loyalty Category
### Step 6 : Create a Spending Flag
### Step 7: Create an Order Frequency Flag
### Step 9: Export Updated dataframe


## Step 1 : Import libraries and dataframe

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Import "ords_prods_merge_derived" with last update 
path= r'C:\Users\Bilel\Desktop\data_analyst_Carrer Foundry\Data Immersion\07-12-24 Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_derived.pkl'))

In [3]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices    price_range_loc     busiest_day Busiest_days  \
0              7     9.0  Mid-range product  Regularly busy  Regular day   
1             16    12.5  Mid-range product  Regularly busy  Regular day   
2             19     4.4  Low-range product  Regularly busy  Regular day   
3             19     4.7  Low-range product  Regularly busy  Regular day   
4             17     1.0  Low-range product  Regularly busy  Regular day   

  busiest_period_of_day  max_order  loyalty_flag  
0        Average orders         10  New customer  
1        Average orders         10  New customer  
2        Average orders         10  New customer  
3        Average orders         10  New customer  
4        Average orders         10  New customer

## Step 2 : Aggregate mean of Order Number by Department

In [4]:
ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Step 3 : How do the results for the entire dataframe differ from those of the subset? 

The differences in the results of the entire dataframe compared to the subset are mainly due to the sample size and potential biases in sampling. In the complete dataframe, the averages of orders per department are calculated from a larger amount of data, allowing for a more accurate and representative estimation. On the other hand, the subset, limited to 1,000,000 rows, may not capture the full diversity of purchasing behaviors and might introduce biases, especially if it is not randomly selected. For example, the average order number for Department 10  is close in both subsets (20.197148 in the entire dataframe and 20.091818 in the subset), indicating stable and well-represented ordering patterns. Meanwhile, Department 16  shows a slight variation, with an average of 17.665606 in the entire dataframe and 18.014071 in the subset. This illustrates how subsets can sometimes deviate from the overall trends observed in the complete dataframe, highlighting the importance of using representative samples to obtain accurate results and avoid biased conclusions.

## Step 4 : Create a Loyalty Flag

This step was already completed on the lesson, first i used trasnform() function which will create a new column containing the maximum frequency of the “order_number” column, and then using the loc() function, which will create a second column containing a flag designating whether a customer is “loyal” or no based to those criteria: 
If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [5]:
# Aggregating Data with transform() Function

#ords_prods_merge['max_order']=ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [6]:
# Creating a 'Loyalty' Label Flag using loc Function

#ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
#ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
#ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [7]:
# Examining Value Counts on 'loyalty_flag'
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [19]:
# Examining Value Counts on 'loyalty_flag' for each user
ords_prods_merge.groupby(['loyalty_flag']).user_id.nunique()

loyalty_flag
Loyal customer       17017
New customer        112328
Regular customer     76864
Name: user_id, dtype: int64

ords_prods_merge['loyalty_flag'].value_counts(dropna = False) here I find the count of each category: Loyal Customer, Regular Customer, and New Customer.. so if a user_id is repeated 20 times, for example, it will be counted 20 times. Logically, of course, we do not have almost 16 million (15876776) Regular customers, almost 10 million (10284093) Loyal customers, and almost 6 million (6243990) New customers.
That's why we should count unique (Distinct) values of user_id in each group: ords_prods_merge.groupby(['loyalty_flag']).user_id.nunique() here each user is counted only once.

In [8]:
# Checking the Updated Output
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   

    days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                      NaN          True         196                  1   
1                      NaN          True       14084                  2   
2                      NaN          True       12427                  3   
3                      NaN          True       26088                  4   
4                      NaN          True       26405                  5   
5                     15.0         False         196                  1   
6                     15.0         False       10258                  2   
7                     15.0         False       12427                  3   
8                     15.0         False       13176                  4   
9                     15.0         False       26088                  5   
10                    15.0         False       13032                  6   
11                    21.0         False         196                  1   
12                    21.0         False       12427                  2   
13                    21.0         False       10258                  3   
14                    21.0         False       25133                  4   

    reordered                             product_name  aisle_id  \
0           0                                     Soda        77   
1           0  Organic Unsweetened Vanilla Almond Milk        91   
2           0                      Original Beef Jerky        23   
3           0               Aged White Cheddar Popcorn        23   
4           0         XL Pick-A-Size Paper Towel Rolls        54   
5           1                                     Soda        77   
6           0                               Pistachios       117   
7           1                      Original Beef Jerky        23   
8           0                   Bag of Organic Bananas        24   
9           1               Aged White Cheddar Popcorn        23   
10          0                    Cinnamon Toast Crunch       121   
11          1                                     Soda        77   
12          1                      Original Beef Jerky        23   
13          1                               Pistachios       117   
14          0                    Organic String Cheese        21   

    department_id  prices    price_range_loc     busiest_day Busiest_days  \
0               7     9.0  Mid-range product  Regularly busy  Regular day   
1              16    12.5  Mid-range product  Regularly busy  Regular day   
2              19     4.4  Low-range product  Regularly busy  Regular day   
3              19     4.7  Low-range product  Regularly busy  Regular day   
4              17     1.0  Low-range product  Regularly busy  Regular day   
5  

## Step 5 : Analyze Spending Habits by Loyalty Category

In [9]:
# Determining Basic Statistics of Spending Habits by Loyalty Category
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Loyal customers purchase products at lower average prices (10.39) compared to regular customers (12.50) and new customers (13.29), which might indicate that they seek and buy more cheaper products or benefit from special discounts. The minimum price is constant at 1.0 for all customer categories, while the maximum price of 99999.0 suggests a possible anomaly in the data.

## Step 6 : Create a Spending Flag

In [10]:
# Aggregating Data with transform() Function
ords_prods_merge['avg_prods_price']= ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Bilel\AppData\Local\Temp\ipykernel_17028\3299760277.py:2: FutureWarning: The provided callable <function mean at 0x000001FD65AACF40> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_prods_price']= ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [11]:
#Creating a "spending_flag" using loc Function
ords_prods_merge.loc[ords_prods_merge['avg_prods_price'] >= 10, 'spending_flag'] = 'High spender'
ords_prods_merge.loc[ords_prods_merge['avg_prods_price'] <  10, 'spending_flag'] = 'Low spender'



In [12]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [20]:
# Examining Value Counts on 'spending_flag' for each user
ords_prods_merge.groupby(['spending_flag']).user_id.nunique()

spending_flag
High spender      5357
Low spender     200852
Name: user_id, dtype: int64

## Step 7: Create an Order Frequency Flag

In [13]:
# Aggregating Data with transform() Function
ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Bilel\AppData\Local\Temp\ipykernel_17028\305945996.py:2: FutureWarning: The provided callable <function median at 0x000001FD65CC7BA0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [14]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ... prices    price_range_loc     busiest_day  Busiest_days  \
0          0  ...    9.0  Mid-range product  Regularly busy   Regular day   
1          0  ...   12.5  Mid-range product  Regularly busy   Regular day   
2          0  ...    4.4  Low-range product  Regularly busy   Regular day   
3          0  ...    4.7  Low-range product  Regularly busy   Regular day   
4          0  ...    1.0  Low-range product  Regularly busy   Regular day   

  busiest_period_of_day max_order  loyalty_flag avg_prods_price  \
0        Average orders        10  New customer        6.367797   
1        Average orders        10  New customer        6.367797   
2        Average orders        10  New customer        6.367797   
3        Average orders        10  New customer        6.367797   
4        Average orders        10  New customer        6.367797   

   spending_flag customer_frequency  
0    Low spender               20.5  
1    Low spender               20.5  
2    Low spender               20.5  
3    Low spender               20.5  
4    Low spender               20.5  

[5 rows x 23 columns]

In [15]:
# Creating flag on order frequency
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] > 20, 'order_frequency_flag'] = 'Non_frequent customer'
ords_prods_merge.loc[(ords_prods_merge['customer_frequency']>10) & (ords_prods_merge['customer_frequency']<=20),'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent Customer'

In [16]:
# Examining Value Counts on 'order_frequency_flag'
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent Customer        21559853
Regular customer          7208564
Non_frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [21]:
# Examining Value Counts on 'spending_flag' for each user
ords_prods_merge.groupby(['order_frequency_flag']).user_id.nunique()

order_frequency_flag
Frequent Customer        86596
Non_frequent customer    59619
Regular customer         59993
Name: user_id, dtype: int64

In [17]:
# Checking Relevant Columns of Updated Dataframe
ords_prods_merge[['user_id', 'days_since_prior_order', 'customer_frequency', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  customer_frequency   order_frequency_flag
0         1                     NaN                20.5  Non_frequent customer
1         1                     NaN                20.5  Non_frequent customer
2         1                     NaN                20.5  Non_frequent customer
3         1                     NaN                20.5  Non_frequent customer
4         1                     NaN                20.5  Non_frequent customer
5         1                    15.0                20.5  Non_frequent customer
6         1                    15.0                20.5  Non_frequent customer
7         1                    15.0                20.5  Non_frequent customer
8         1                    15.0                20.5  Non_frequent customer
9         1                    15.0                20.5  Non_frequent customer
10        1                    15.0                20.5  Non_frequent customer
11        1                    21.0                20.5  Non_frequent customer
12        1                    21.0                20.5  Non_frequent customer
13        1                    21.0                20.5  Non_frequent customer
14        1                    21.0                20.5  Non_frequent customer

## Step 9: Export Updated dataframe

In [18]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_grouped.pkl'))